2020-04-introducing-mdjs-live-demos-everywhere-local-github-storybook/README.md

```diff
@@ -0,0 +1,344 @@
+
+## Introducing Markdown with JavaScript (mdjs)
+
+The primary goals are
+
+- minimal complexity
+- follows progressive enhancement
+- markdown needs to still make sense
+- code highlighting in editors without additional tools
+- looks good on github/gitlab/any source code management tool
+
+The fundamental idea is laughable simple. We "enhance" a code fence block with additional meta data `js script`.
+
+    ```js script
+    import './my-component.js';
+    ```
+    # This is my component
+    <my-component></my-component>
+
+And that's it 😅
+
+Enough talk you can see it here live:
+
+[==> Link to editable demo <==](https://webcomponents.dev/edit/collection/lsZ2eaviDNwy6pIBEDeL/aPQdZ4FtAiqJ7YXnRe2s)
+
+**How does it work**
+
+Mdjs hooks into [remark](https://remark.js.org/) and extracts all those tagged js blocks.
+In the end, html and js is separately available.
+
+```js
+{
+  html: '<h1>This is my component</h1><my-component></my-component>',
+  jsCode: "import './my-component.js';"
+}
+```
+
+It can then be combined/processed by any tool to create an actual documentation page.
+
+The process looks like this:
+
+1. Extract `js script` and separate it from md
+2. Render md
+3. Provide html & js
+
+==> INSERT ANIMATION <==
+
+This already is powerful enough to directly include JavaScript and render web components with attributes.
+
+## Enhancing mdjs with demo format
+
+Now that we can execute JavaScript within our Markdown this opens the door for more advanced features.
+
+Our first step is to create another enhanced js code block `js preview-story`.
+In there you should export a function that can be executed on demand.
+This will add a border around and a button to show/hide the actual source code.
+
+    ```js script
+    import './my-component.js';
+    ```
+    # This is my component
+    ```js preview-story
+    export const demo = () => `<my-component header="from attribute"></my-component>`
+    ```
+
+> if you do not want to border and button you can use `js story`
+
+What you get looks something like this
+
+```js
+{
+  html: '<h1>This is my component</h1><my-component></my-component>',
+  jsCode: "import './my-component.js';",
+  stories: [
+    key: 'demo',
+    name: 'demo',
+    code: 'export const demo = () => `<my-component header="from attribute"></my-component>`',
+  ]
+}
+```
+
+This adds an extra step to the processing:
+
+1. Extract `js script` and separate from md
+2. Extract `js story` and `js preview-story` and separate from md
+3. Put a placeholder `<mdjs-story mdjs-story-name="demo"></mdjs-story>` or `mdjs-preview` at it's place
+4. Render md
+5. Provide html & js & stories
+
+With that information you can now create full javascript and demo capable pages purely from markdown.
+
+By default Mdjs takes it a small step further by supporting an actual template system - namely [lit-html](https://lit-html.polymer-project.org/).
```

Comment:
I don't know if you want to expose this support as part of `mdjs`. You can support this because `@mdjs/story(preview)` is written in LitElement. If you write that it means `mdjs` must support lit-html as a spec... There is a subtle line here.

Reply:
This gives me an idea to uncouple `@mdjs/core` and `@mdjs/preview` / `@mdjs/stories` and the lit-html dependency.

```js
```js story use <mdjsstory-lit>
export const demo = () => html`
      <my-component .header=${name}></my-component>
    `;
```
```

this tells, render this story with `<mdjs-story-lit>` tag.

You simply add `import 'mdjsstory-lit';` in `js script` on the top.

This way `@mdjs/core` supports "stories" but the rendering is completely plugable.

my 2 cents

Reply:
will discuss in https://github.com/open-wc/open-wc/issues/1498